Dockerfile

```diff
@@ -1,34 +1,77 @@
-
+# Use UTF-8.
 RUN locale-gen en_US.UTF-8
 ENV LANG en_US.UTF-8
 ENV LANGUAGE en_US:en
 ENV LC_ALL en_US.UTF-8
-ENV ANALYTICS_API_CFG /edx/etc/analytics_api.yml
 
-WORKDIR /edx/app/analytics_api
-COPY requirements /edx/app/analytics_api/requirements
-RUN pip install -r requirements/production.txt
+ARG COMMON_APP_DIR="/edx/app"
+ARG ANALYTICS_API_SERVICE_NAME="analytics_api"
+ENV ANALYTICS_API_HOME "${COMMON_APP_DIR}/${ANALYTICS_API_SERVICE_NAME}"
+ARG ANALYTICS_API_APP_DIR="${COMMON_APP_DIR}/${ANALYTICS_API_SERVICE_NAME}"
+ARG ANALYTICS_API_VENV_DIR="${COMMON_APP_DIR}/${ANALYTICS_API_SERVICE_NAME}/venvs/${ANALYTICS_API_SERVICE_NAME}"
+ARG ANALYTICS_API_CODE_DIR="${ANALYTICS_API_APP_DIR}/${ANALYTICS_API_SERVICE_NAME}"
+
+ENV ANALYTICS_API_CODE_DIR="${ANALYTICS_API_CODE_DIR}"
+ENV PATH "${ANALYTICS_API_VENV_DIR}/bin:$PATH"
+ENV COMMON_CFG_DIR "/edx/etc"
+ENV ANALYTICS_API_CFG "/edx/etc/${ANALYTICS_API_SERVICE_NAME}.yml"
+
+# Working directory will be root of repo.
+WORKDIR ${ANALYTICS_API_CODE_DIR}
+
+RUN virtualenv -p python3.8 --always-copy ${ANALYTICS_API_VENV_DIR}
+
+# Expose canonical Analytics port
+EXPOSE 19001
+
+FROM base as prod
+
+ENV DJANGO_SETTINGS_MODULE "analyticsdataserver.settings.production"
+
+COPY requirements/production.txt ${ANALYTICS_API_CODE_DIR}/requirements/production.txt
+
+RUN pip install -r ${ANALYTICS_API_CODE_DIR}/requirements/production.txt
+
+# Copy over rest of code.
+# We do this AFTER requirements so that the requirements cache isn't busted
+# every time any bit of code is changed.
+
+COPY . .
+
+# exec /edx/app/analytics_api/venvs/analytics_api/bin/gunicorn -c /edx/app/analytics_api/analytics_api_gunicorn.py  analyticsdataserver.wsgi:application
+
+CMD ["gunicorn" , "-b", "0.0.0.0:8100", "--pythonpath", "/edx/app/analytics_api/analytics_api","analyticsdataserver.wsgi:application"]
+
+FROM base as dev
+
+ENV DJANGO_SETTINGS_MODULE "analyticsdataserver.settings.devstack"
+
+COPY requirements/dev.txt ${ANALYTICS_API_CODE_DIR}/requirements/dev.txt
+
+RUN pip install -r ${ANALYTICS_API_CODE_DIR}/requirements/dev.txt
 
-EXPOSE 8100
-CMD gunicorn --bind=0.0.0.0:8100 --workers 2 --max-requests=1000 -c /edx/app/analytics_api/analytics_data_api/docker_gunicorn_configuration.py analyticsdataserver.wsgi:application
+# Copy over rest of code.
+# We do this AFTER requirements so that the requirements cache isn't busted
+# every time any bit of code is changed.
+COPY . .
 
-RUN useradd -m --shell /bin/false app
-USER app
-COPY . /edx/app/analytics_api
+# Devstack related step for backwards compatibility
+RUN touch /edx/app/${ANALYTICS_API_SERVICE_NAME}/${ANALYTICS_API_SERVICE_NAME}_env
 
-FROM app as newrelic
-RUN pip install newrelic
-CMD newrelic-admin run-program gunicorn --bind=0.0.0.0:8100 --workers 2 --max-requests=1000 -c /edx/app/analytics_api/analytics_data_api/docker_gunicorn_configuration.py analyticsdataserver.wsgi:application
+CMD while true; do python ./manage.py runserver 0.0.0.0:8110; sleep 2; done
```

Comment:
We are no longer hooking up with newrelic-admin with this change. Why would you want to leave that out?

Reply:
We don't use new relic in `devstack`. As for the current case I haven't added it to prod target as well since according to our discussion with Jeremy we don't want to make `new relic` a necessary part of our base images by default. We can make a separate target for that in our Dockerfile or leave it to our orchestration system(k8s or docker-compose) to do those things for us.

Reply:
Got it. Do you have wiki page documents talk about this? I'd like this point to be written down as a policy for layering newrelic to our apps.